### PR TITLE
PARQUET-490: [C++][Parquet] Basic support for reading DELTA_BINARY_PACKED data

### PR DESCRIPTION
**Pull request by shanhuuang**
Only works for int32/int64 when delta_bit_width_<=32.

1. Update the implement of DeltaBitPackDecoder
2. Revise the code of putting/getting ZigZag int and the unit test
3. Add the implement of putting/getting ZigZag int64 and unit test

### CODE REVIEW

File: cpp/src/arrow/util/bit_stream_utils.h [hunk 1]
```diff
@@ -418,14 +450,65 @@ inline bool BitReader::GetVlqInt(uint32_t* v) {
 }
 
 inline bool BitWriter::PutZigZagVlqInt(int32_t v) {
-  auto u_v = ::arrow::util::SafeCopy<uint32_t>(v);
-  return PutVlqInt((u_v << 1) ^ (u_v >> 31));
+  uint32_t u_v = ::arrow::util::SafeCopy<uint32_t>(v);
+  v = (u_v << 1) ^ (v >> 31);
+  u_v = ::arrow::util::SafeCopy<uint32_t>(v);
```
Comment (emkornfield) [inline cpp/src/arrow/util/bit_stream_utils.h:455]: nit: just static_cast v >> 31 above.

Comment (emkornfield) [reply cpp/src/arrow/util/bit_stream_utils.h:455]: and elimnate this line.

File: cpp/src/arrow/util/bit_stream_utils.h [hunk 2]
```diff
@@ -418,14 +450,65 @@ inline bool BitReader::GetVlqInt(uint32_t* v) {
...
+  return PutVlqInt(u_v);
 }
 
 inline bool BitReader::GetZigZagVlqInt(int32_t* v) {
   uint32_t u;
   if (!GetVlqInt(&u)) return false;
-  *v = ::arrow::util::SafeCopy<int32_t>((u >> 1) ^ (u << 31));
+  *v = ::arrow::util::SafeCopy<int32_t>(u);
```
Comment (emkornfield) [inline cpp/src/arrow/util/bit_stream_utils.h:462]: https://github.com/protocolbuffers/protobuf/blob/0b05994474f3f53bbda8fef1c24acb06f778a3bf/src/google/protobuf/wire_format_lite.h#L867 seems to be more concise and has a comptable license, could we use that and update license.txt to note the use of the code.

File: cpp/src/arrow/util/bit_stream_utils.h [hunk 3]
```diff
@@ -263,8 +284,10 @@ inline void GetValue_(int num_bits, T* v, int max_bytes, const uint8_t* buffer,
 #pragma warning(disable : 4800 4805)
 #endif
     // Read bits of v that crossed into new buffered_values_
-    *v = *v | static_cast<T>(BitUtil::TrailingBits(*buffered_values, *bit_offset)
-                             << (num_bits - *bit_offset));
+    if (ARROW_PREDICT_TRUE(num_bits - *bit_offset < 64)) {
```
Comment (emkornfield) [inline cpp/src/arrow/util/bit_stream_utils.h:287]: please comment why this check is needed.

File: cpp/src/arrow/util/bpacking.cc [hunk 4]
```diff
@@ -174,5 +176,221 @@ int unpack32(const uint32_t* in, uint32_t* out, int batch_size, int num_bits) {
 #endif
 }
 
+namespace {
+
+int unpack64_default(const uint8_t* in, uint64_t* out, int batch_size, int num_bits) {
```
Comment (emkornfield) [inline cpp/src/arrow/util/bpacking.cc:181]: please comment on how this code was generated?  If there is a script could you add that as well.

File: cpp/src/arrow/util/bpacking64_default.h [hunk 5]
```diff
@@ -0,0 +1,5642 @@
...
+
+#pragma once
+
+#include "arrow/util/bit_util.h"
+#include "arrow/util/ubsan.h"
+
+namespace arrow {
+namespace internal {
+
+inline const uint8_t* unpack0_64(const uint8_t* in, uint64_t* out) {
```
Comment (emkornfield) [inline cpp/src/arrow/util/bpacking64_default.h:35]: same comment as above, if there is a script used to generate this please add it to the PR.

File: cpp/src/parquet/encoding.cc [hunk 6]
```diff
@@ -2062,17 +2061,12 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
 
   explicit DeltaBitPackDecoder(const ColumnDescriptor* descr,
                                MemoryPool* pool = ::arrow::default_memory_pool())
-      : DecoderImpl(descr, Encoding::DELTA_BINARY_PACKED), pool_(pool) {
-    if (DType::type_num != Type::INT32 && DType::type_num != Type::INT64) {
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:2066]: why remove this check?

File: cpp/src/parquet/reader_test.cc [hunk 7]
```diff
@@ -322,6 +323,27 @@ TEST(TestFileReaderAdHoc, NationDictTruncatedDataPage) {
...
+  std::list<int> columns;
+
+  std::stringstream ss_values;
+  const char* file = "delta_binary_packed.parquet";
+  auto reader_props = default_reader_properties();
+  auto reader = ParquetFileReader::OpenFile(data_file(file), false, reader_props);
+  ParquetFilePrinter printer(reader.get());
+  printer.DebugPrint(ss_values, columns, true, false, false, file);
+
+  std::ifstream in(data_file("delta_binary_packed.parquet.expect"));
```
Comment (emkornfield) [inline cpp/src/parquet/reader_test.cc:340]: this test seems very brittle.  using CSV or JSON or some other format and comparing the column data seems directly instead of the string representation seems like a better idea?

**Comment by emkornfield** [comment]
Is it possible for delta_bit_width_<=32?  if so it seems like we should probably cover that case while updating?

**Comment by emkornfield** [comment]
CC @zeroshade note bugs fixed in ZigZag/VLQ int, not sure if these got propagated to Go.

**Comment by zeroshade** [comment]
@emkornfield I added the tests that were added here to the Go side, including the comparison of the actual bytes generated and the Golang implementation passes all the tests as currently written without any modifications needed. it appears that the Golang version doesn't have this bug (I remember coming across a bug in the ZigZag/VLQ int stuff a while back where I had to manage the signs and such to get the right output so I likely fixed it without realizing).

File: cpp/src/arrow/util/bit_stream_utils.h (hunk 3 above)
Comment (shanhuuang) [reply cpp/src/arrow/util/bit_stream_utils.h:287]: To solve the runtime error in test job "AMD64 Ubuntu 20.04 C++ ASAN UBSAN":
/arrow/cpp/src/arrow/util/bit_stream_utils.h:289:30: runtime error: shift exponent 64 is too large for 64-bit type 'uint64_t' (aka 'unsigned long')
https://github.com/shanhuuang/arrow/runs/2986826720?check_suite_focus=true#logs
I met it when I try to read a file with only 3 records.

File: cpp/src/arrow/util/bit_stream_utils.h (hunk 1 above)
Comment (shanhuuang) [reply cpp/src/arrow/util/bit_stream_utils.h:455]: Actually, I'm not particularly clear about the difference between ::arrow::util::SafeCopy<uint32_t> and static_cast<uint32_t>. Could you please tell me when should we use SafeCopy rather than static_cast?

File: cpp/src/parquet/encoding.cc (hunk 6 above)
Comment (shanhuuang) [reply cpp/src/parquet/encoding.cc:2066]: I thought that DeltaBitPackDecoder object would only be created in function parquet::MakeDecoder, so no more check is needed here. It's ok to restore this code.

**Comment by shanhuuang** [comment]
> Is it possible for delta_bit_width_<=32? if so it seems like we should probably cover that case while updating?

In current code, internal::unpack32 is called in BitReader::GetBatch if delta_bit_width_<=32 whether value type is int64 or int32.
delta_binary_packed.parquet contains int64 columns with delta_bit_width_ from 0 to 32

File: LICENSE.txt [hunk 8]
```diff
@@ -2240,3 +2241,46 @@ be copied, modified, or distributed except according to those terms.
 
 Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 express or implied.  See your chosen license for details.
+
+--------------------------------------------------------------------------------
+
+The file cpp/src/arrow/util/bit_stream_utils.h contains code from
```
Comment (emkornfield) [inline LICENSE.txt:2247]: there is already a reference to protocol buggers above, could you update that instead?

**Comment by emkornfield** [comment]
> > Is it possible for delta_bit_width_<=32? if so it seems like we should probably cover that case while updating?
> 
> In current code, internal::unpack32 is called in BitReader::GetBatch if delta_bit_width_<=32 whether value type is int64 or int32.
> delta_binary_packed.parquet contains int64 columns with delta_bit_width_ from 0 to 32

Sorry I think i mean >= 32?

File: cpp/src/arrow/util/bit_stream_utils.h (hunk 1 above)
Comment (emkornfield) [reply cpp/src/arrow/util/bit_stream_utils.h]: Using safe-copy instead could also work.  SafeCopy (and other methods in ubsan.h) are  mostly to avoid undefined behavior.  The most common case of this is when trying to read possibly unaligned values.  

SafeCopy is equivelant to the new [std::bit_cast](https://en.cppreference.com/w/cpp/numeric/bit_cast)

**Comment by shanhuuang** [comment]
> > > Is it possible for delta_bit_width_<=32? if so it seems like we should probably cover that case while updating?
> > 
> > 
> > In current code, internal::unpack32 is called in BitReader::GetBatch if delta_bit_width_<=32 whether value type is int64 or int32.
> > delta_binary_packed.parquet contains int64 columns with delta_bit_width_ from 0 to 32
> 
> Sorry I think i mean >= 32?

Yes, it is possible for delta_bit_width_>32. For example, a column contains records like 0, -4294967296, -860510501
The deltas would be -4294967296 and 3434456795
The minimum delta is -4294967296
so the final deltas are 0 and 7729424091
delta_bit_width_ of bit packed int(7729424091) is 33

delta_binary_packed.parquet also contains int64 columns with delta_bit_width_ from 33 to 64.

**Comment by shanhuuang** [comment]
@emkornfield Could you please review it? Thank you very much!

**Comment by emkornfield** [comment]
Sorry will try to get to it next week.

**Comment by shanhuuang** [comment]
> Sorry will try to get to it next week.

Sorry for the delay in getting back to you. This week would be fine :)
I'm also working on PARQUET-491 and PARQUET-492 which depend on code in this PR. Maybe I should create a new PR after this one is merged.

**Comment by emkornfield** [comment]
sorry I have had less time then I would have liked recently for Arrow reviews, will try to get to this soon.

**Comment by shanhuuang** [comment]
> sorry I have had less time then I would have liked recently for Arrow reviews, will try to get to this soon.

OK, Thanks a lot :)

File: cpp/src/arrow/util/bit_stream_utils.h (hunk 3 above)
Comment (emkornfield) [reply cpp/src/arrow/util/bit_stream_utils.h:287]: thanks, makes esnse, could add a comment to the code.

File: cpp/src/arrow/util/bit_stream_utils.h [hunk 9]
```diff
@@ -418,14 +450,59 @@ inline bool BitReader::GetVlqInt(uint32_t* v) {
...
-  *v = ::arrow::util::SafeCopy<int32_t>((u >> 1) ^ (u << 31));
+  u = (u >> 1) ^ (~(u & 1) + 1);
+  *v = ::arrow::util::SafeCopy<int32_t>(u);
+  return true;
+}
+
+inline bool BitWriter::PutVlqInt(uint64_t v) {
+  bool result = true;
+  while ((v & 0xFFFFFFFFFFFFFF80ULL) != 0ULL) {
+    result &= PutAligned<uint8_t>(static_cast<uint8_t>((v & 0x7F) | 0x80), 1);
```
Comment (emkornfield) [inline cpp/src/arrow/util/bit_stream_utils.h:469]: nit: should this be logical and?

File: cpp/src/parquet/encoding.cc [hunk 10]
```diff
@@ -2107,73 +2105,99 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
...
 
-      // TODO: the key to this algorithm is to decode the entire miniblock at once.
-      int64_t delta;
-      if (!decoder_.GetValue(delta_bit_width_, &delta)) ParquetException::EofException();
-      delta += min_delta_;
-      last_value_ += static_cast<int32_t>(delta);
-      buffer[i] = last_value_;
-      --values_current_mini_block_;
+      int values_decode =
+          std::min(values_current_mini_block_, (uint32_t)(max_values - i));
```
Comment (emkornfield) [inline cpp/src/parquet/encoding.cc:2166]: please use static_cast.  why the cast to uint?

**Comment by emkornfield** [comment]
Could we create a separate PR for the test files directly to the test file repo?

**Comment by emkornfield** [comment]
I might be missing it but is there a test demonstrating current behavior for bit_widths >= 32?

**Comment by emkornfield** [comment]
Mostly looks OK, I'm going to try to do a second more careful pass this week.  Also CC @pitrou in case he wants to take a look.

**Comment by emkornfield** [comment]
Thank you for your patience.

File: cpp/src/arrow/util/bit_stream_utils.h (hunk 9 above)
Comment (shanhuuang) [reply cpp/src/arrow/util/bit_stream_utils.h:469]: I used "&=" here with reference to the code in function "bool BitWriter::PutVlqInt(uint32_t v)". Using "&&=" would be a better choice.

File: cpp/src/parquet/encoding.cc (hunk 10 above)
Comment (shanhuuang) [reply cpp/src/parquet/encoding.cc:2166]: because the type of values_current_mini_block_ is uint32_t. I'll use static_cast, thanks~

File: cpp/src/parquet/encoding.cc [hunk 11]
```diff
@@ -2107,73 +2105,99 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
...
       ParquetException::EofException();
     }
     if (!decoder_.GetZigZagVlqInt(&last_value_)) ParquetException::EofException();
 
-    delta_bit_widths_ = AllocateBuffer(pool_, num_mini_blocks_);
-    uint8_t* bit_width_data = delta_bit_widths_->mutable_data();
+    delta_bit_widths_ = AllocateBuffer(pool_, mini_blocks_per_block_);
+    values_per_mini_block_ = values_per_block_ / mini_blocks_per_block_;
+    if (values_per_mini_block_ % 8 != 0) {
+      throw ParquetException("miniBlockSize must be multiple of 8, but it's " +
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:2119]: Why 8? According to the [spec](https://github.com/apache/parquet-format/blob/master/Encodings.md#delta-encoding-delta_binary_packed--5), it should be a multiple of 32.

File: cpp/src/parquet/encoding.cc [hunk 12]
```diff
@@ -2107,73 +2105,99 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
...
+
+    mini_block_idx_ = mini_blocks_per_block_;
+    values_current_mini_block_ = 0;
+  }
 
+  void InitBlock() {
     if (!decoder_.GetZigZagVlqInt(&min_delta_)) ParquetException::EofException();
-    for (uint32_t i = 0; i < num_mini_blocks_; ++i) {
+
+    // readBitWidthsForMiniBlocks
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:2130]: What does this comment mean? Would you like to write it in plain English?

File: cpp/src/parquet/encoding.cc [hunk 13]
```diff
@@ -2107,73 +2105,99 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
...
       if (ARROW_PREDICT_FALSE(values_current_mini_block_ == 0)) {
         ++mini_block_idx_;
-        if (mini_block_idx_ < static_cast<size_t>(delta_bit_widths_->size())) {
-          delta_bit_width_ = bit_width_data[mini_block_idx_];
+        if (mini_block_idx_ < mini_blocks_per_block_) {
+          delta_bit_width_ = delta_bit_widths_->data()[mini_block_idx_];
           values_current_mini_block_ = values_per_mini_block_;
         } else {
+          // mini_block_idx_ > mini_blocks_per_block_ only if last_value_ is stored in
+          // header
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:2154]: I don't understand this comment. `last_value_` is always stored in header, right?

File: cpp/src/parquet/encoding.cc [hunk 14]
```diff
@@ -2107,73 +2105,99 @@ class DeltaBitPackDecoder : public DecoderImpl, virtual public TypedDecoder<DTyp
...
+          if (ARROW_PREDICT_FALSE(mini_block_idx_ > mini_blocks_per_block_)) {
```
Comment (pitrou) [inline cpp/src/parquet/encoding.cc:2155]: Instead of this cryptic condition, can you add a boolean member indicating that decoding hasn't started yet?

File: cpp/src/arrow/util/bpacking64_codegen.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        print("  uint64_t w{0} = util::SafeLoadAs<uint64_t>(in);".format(k))
+        print("  w{0} = arrow::BitUtil::FromLittleEndian(w{0});".format(k))
+        print("  in += 8;".format(k))
+    k = howmanywords(bit) - 1
+    if (bit % 2 == 0):
+        print("  uint64_t w{0} = util::SafeLoadAs<uint64_t>(in);".format(k))
+        print("  w{0} = arrow::BitUtil::FromLittleEndian(w{0});".format(k))
+        print("  in += 8;".format(k))
+    else:
+        print("  uint64_t w{0} = util::SafeLoadAs<uint32_t>(in);".format(k))
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/util/bpacking64_codegen.py:103]: Ok, is this because DELTA_BINARY_PACKED outputs values in multiples of 32? Slightly annoying :-/

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for this PR. Are you planning to support encoding in another PR?

**Comment by pitrou** [comment]
> Only works for int32/int64 when delta_bit_width_<=32.

What do you mean? Are delta bit widths > 32 not supported? I don't see check in the code.

**Review by shanhuuang — COMMENTED** [review]
> > Only works for int32/int64 when delta_bit_width_<=32.
> 
> What do you mean? Are delta bit widths > 32 not supported? I don't see check in the code.

Sorry, this is the commit message when the code was submitted in the first time. delta bit widths > 32 have also been supported now in this PR.

**Comment by shanhuuang** [comment]
> Thank you for this PR. Are you planning to support encoding in another PR?

Yes. I'm also working on [PARQUET-491](https://issues.apache.org/jira/browse/PARQUET-491) and [PARQUET-492](https://issues.apache.org/jira/browse/PARQUET-492) which depend on code in this PR. And I would like to create a new PR after this one is merged.

File: cpp/src/parquet/CMakeLists.txt [hunk 16]
```diff
@@ -407,3 +407,7 @@ endif()
 if(ARROW_WITH_ZSTD)
   add_definitions(-DARROW_WITH_ZSTD)
 endif()
+
+if(ARROW_CSV)
```
Comment (emkornfield) [inline cpp/src/parquet/CMakeLists.txt:411]: this seems out of place?

Comment (shanhuuang) [reply cpp/src/parquet/CMakeLists.txt:411]: Do you means that I should add the code in function [ADD_PARQUET_TEST](https://github.com/apache/arrow/blob/master/cpp/src/parquet/CMakeLists.txt#L34)? Something like:
```
  if(REL_TEST_NAME STREQUAL "arrow-test" AND ARROW_CSV)
    add_definitions(-DARROW_CSV)
  endif()
```
This one also works :)

File: cpp/src/parquet/encoding.cc (hunk 11 above)
Comment (pitrou) [reply cpp/src/parquet/encoding.cc:2119]: Well, we should follow the spec here, especially as `unpack64` assumes the input length is a multiple of 32.

Comment (pitrou) [reply cpp/src/parquet/encoding.cc:2119]: Also, perhaps open a JIRA about the parquet-mr `DeltaBinaryPackingConfig` check?

**Review by pitrou — CHANGES_REQUESTED** [review]
This is starting to look quite good. Can you update the `parquet-testing` submodule now that https://github.com/apache/parquet-testing/pull/19 has been merged?

Also, here are a couple more comments.

File: cpp/src/arrow/util/bit_util_test.cc [hunk 17]
```diff
@@ -1939,24 +1939,61 @@ TEST(BitUtil, RoundUpToPowerOf2) {
...
+static void TestZigZag(int32_t v, std::array<uint8_t, 5> buffer_expect) {
   uint8_t buffer[BitUtil::BitReader::kMaxVlqByteLength] = {};
   BitUtil::BitWriter writer(buffer, sizeof(buffer));
   BitUtil::BitReader reader(buffer, sizeof(buffer));
   writer.PutZigZagVlqInt(v);
+  EXPECT_EQ(buffer_expect[0], buffer[0]);
+  EXPECT_EQ(buffer_expect[1], buffer[1]);
+  EXPECT_EQ(buffer_expect[2], buffer[2]);
+  EXPECT_EQ(buffer_expect[3], buffer[3]);
+  EXPECT_EQ(buffer_expect[4], buffer[4]);
```
[the remark this answers is not in the document]
Comment (shanhuuang) [reply cpp/src/arrow/util/bit_util_test.cc]: Got it! Thanks. :)

File: cpp/src/arrow/util/bit_stream_utils.h [hunk 18]
```diff
@@ -313,7 +337,18 @@ inline int BitReader::GetBatch(int num_bits, T* v, int batch_size) {
                            reinterpret_cast<uint32_t*>(v + i), batch_size - i, num_bits);
     i += num_unpacked;
     byte_offset += num_unpacked * num_bits / 8;
+  } else if (sizeof(T) == 8 && num_bits > 32) {
+    // Use unpack64 only if num_bits is larger then 32
```
[the remark this answers is not in the document]
Comment (shanhuuang) [reply cpp/src/arrow/util/bit_stream_utils.h]: Yes, I think it is just a performance issue. If num_bits is no larger than 32, I guess that using unpack32 will achieve better performance with function such as unpack32_avx2/unpack32_avx512. 
The result of using unpack32 or unpack64 are the same if num_bits <= 32

File: cpp/src/parquet/encoding.cc (hunk 11 above)
Comment (shanhuuang) [reply cpp/src/parquet/encoding.cc]: OK. I've opened a JIRA [here](https://issues.apache.org/jira/browse/PARQUET-2077).

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@shanhuuang Do you have a user id on the [Apache JIRA tracker](https://issues.apache.org/jira/), so that I can assign the issue to you?

**Comment by shanhuuang** [comment]
> @shanhuuang Do you have a user id on the [Apache JIRA tracker](https://issues.apache.org/jira/), so that I can assign the issue to you?

Yes, my username on JIRA is "shanhuang". Thanks! :)
